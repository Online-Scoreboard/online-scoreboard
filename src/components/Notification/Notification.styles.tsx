import { makeStyles } from '@material-ui/core/styles';
import { amber, green } from '@material-ui/core/colors';

export const useStyles = makeStyles(theme => ({
  success: {
    backgroundColor: green[600],
    color: theme.palette.getContrastText(green[600]),
  },
  error: {
    backgroundColor: theme.palette.error.dark,
    color: theme.palette.error.contrastText,
  },
  info: {
    backgroundColor: theme.palette.primary.main,
    color: theme.palette.primary.contrastText,
  },
  warning: {
    backgroundColor: amber[700],
    color: theme.palette.getContrastText(amber[700]),
  },
  icon: {
    fontSize: 20,
  },
  iconVariant: {
    fontSize: 20,
    opacity: 0.9,
    marginRight: theme.spacing(1),
  },
  message: {
    display: 'flex',
    alignItems: 'center',
  },
}));
