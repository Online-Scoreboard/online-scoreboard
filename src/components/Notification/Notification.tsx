import React, { memo, useMemo } from 'react';
import { Snackbar, SnackbarContent } from '@material-ui/core';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import ErrorIcon from '@material-ui/icons/Error';
import WarningIcon from '@material-ui/icons/Warning';
import InfoIcon from '@material-ui/icons/Info';
import { useStyles } from './Notification.styles';
import withWidth, { WithWidth, isWidthDown } from '@material-ui/core/withWidth';

const variantIcon = {
  success: CheckCircleIcon,
  warning: WarningIcon,
  error: ErrorIcon,
  info: InfoIcon,
};

interface SnackbarContentWrapperProps {
  message: string;
  variant: 'error' | 'info' | 'success' | 'warning';
}

const NotificationContent: React.FC<SnackbarContentWrapperProps> = ({ message, variant }) => {
  const classes = useStyles();
  const Icon = variantIcon[variant];

  return (
    <SnackbarContent
      className={classes[variant]}
      aria-describedby="client-snackbar"
      message={
        <span id="client-snackbar" className={classes.message}>
          <Icon className={classes.iconVariant} />
          {message}
        </span>
      }
    />
  );
};

export type NotificationVariant = 'error' | 'info' | 'success' | 'warning';

interface NotificationProps extends WithWidth {
  message: string | null | undefined;
  variant: NotificationVariant;
  open: boolean;
  handleClose?: () => void;
}

const NotificationComponent: React.FC<NotificationProps> = ({ message, variant, open, width }) => {
  const isMobile = useMemo(() => isWidthDown('sm', width), [width]);

  if (!message) {
    return null;
  }

  return (
    <Snackbar open={open} anchorOrigin={{ vertical: isMobile ? 'bottom' : 'top', horizontal: 'center' }}>
      <NotificationContent message={message} variant={variant} />
    </Snackbar>
  );
};

export const Notification = memo(withWidth()(NotificationComponent));
