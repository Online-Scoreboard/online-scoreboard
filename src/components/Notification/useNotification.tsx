import { useEffect, useRef } from 'react';
import { NotificationVariant } from './Notification';
import { useQuery, useMutation, useApolloClient } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import { Resolvers } from 'apollo-boost';

interface SetOpenInput {
  openStatus: boolean;
}

interface SetVariantInput {
  variant: NotificationVariant;
}

interface SetMessageInput {
  message: string;
}

interface NotificationData {
  notification: {
    open: boolean;
    variant: NotificationVariant;
    message: string;
    __typename: string;
  };
}

export const notificationInitialData: NotificationData = {
  notification: {
    __typename: 'Notification',
    message: '',
    variant: 'info',
    open: false,
  },
};

const GET_NOTIFICATION = gql`
  query GetNotification {
    notification @client {
      message
      variant
      open
    }
  }
`;

const SET_OPEN = gql`
  mutation SetOpen($openStatus: Boolean!) {
    setOpen(openStatus: $openStatus) @client
  }
`;

const SET_MESSAGE = gql`
  mutation SetMessage($message: String!) {
    setMessage(message: $message) @client
  }
`;

const SET_VARIANT = gql`
  mutation SetVariant($variant: Variant!) {
    setVariant(variant: $variant) @client
  }
`;

const resolvers: Resolvers = {
  Mutation: {
    async setOpen(_, { openStatus }: SetOpenInput, { cache, getCacheKey }) {
      const currState = cache.readQuery({ query: GET_NOTIFICATION });

      cache.writeData({
        data: {
          notification: {
            ...currState.notification,
            open: openStatus,
          },
        },
      });

      return null;
    },

    async setMessage(_, { message }: SetMessageInput, { cache, getCacheKey }) {
      const currState = cache.readQuery({ query: GET_NOTIFICATION });

      cache.writeData({
        data: {
          notification: {
            ...currState.notification,
            message: message || '',
          },
        },
      });

      return null;
    },

    async setVariant(_, { variant }: SetVariantInput, { cache, getCacheKey }) {
      const currState = cache.readQuery({ query: GET_NOTIFICATION });

      cache.writeData({
        data: {
          notification: {
            ...currState.notification,
            variant: variant || 'info',
          },
        },
      });

      return null;
    },
  },
};

export const useNotification = (timeout: number = 3000) => {
  const client = useApolloClient();
  client.addResolvers(resolvers);

  const { data } = useQuery<NotificationData>(GET_NOTIFICATION);
  const [_setOpen] = useMutation(SET_OPEN);
  const [_setMessage] = useMutation(SET_MESSAGE);
  const [_setVariant] = useMutation(SET_VARIANT);
  let notificationTimeout = useRef<void | NodeJS.Timeout>();

  const defaultNotification: NotificationData = {
    notification: {
      __typename: 'Notification',
      message: '',
      variant: 'info',
      open: false,
    },
  };

  const notification = (data && data.notification) || defaultNotification.notification;
  const { open, message, variant } = notification;

  useEffect(() => {
    if (open && !notificationTimeout.current) {
      notificationTimeout.current = setTimeout(() => {
        _setOpen({ variables: { openStatus: false } });
      }, timeout);
    }
  }, [open, timeout, _setOpen]);

  useEffect(() => {
    return () => {
      if (notificationTimeout.current) {
        clearTimeout(notificationTimeout.current);
      }
    };
  }, []);

  return {
    open,
    message,
    variant,
    setOpen: (openStatus: boolean) => {
      if (notificationTimeout.current) {
        clearTimeout(notificationTimeout.current);
        notificationTimeout.current = undefined;
      }

      _setOpen({ variables: { openStatus } });
    },
    setMessage: (message: string) => {
      return _setMessage({ variables: { message } });
    },
    setVariant: (variant: NotificationVariant) => {
      _setVariant({ variables: { variant } });
    },
  };
};
