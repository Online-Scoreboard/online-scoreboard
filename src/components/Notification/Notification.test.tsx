import React from 'react';
import { mount, shallow } from 'enzyme';

import { Notification } from './Notification';

describe('Notification', () => {
  it('should render a notification component', () => {
    // Arrange
    const el = (
      <div>
        <Notification message="Allo" open={false} variant="info" />
      </div>
    );

    // Act
    const wrapper = shallow(el);

    // Assert
    expect(wrapper.find(Notification).exists()).toBe(true);
  });

  it('should render an empty notification component when not opened', () => {
    // Arrange
    const el = <Notification message="Allo" open={false} variant="info" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find(Notification).html()).toBe('');
  });

  it('should render a notification component when opened', () => {
    // Arrange
    (global as any).matchMedia = (media: any) => ({
      addListener: () => {},
      removeListener: () => {},
      matches: true,
    });

    const el = <Notification message="Allo" open={true} variant="info" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find(Notification).html()).toMatchSnapshot();
    expect(wrapper.find('span#client-snackbar').text()).toEqual('Allo');
  });

  it('should not render a notification component when the message is empty', () => {
    // Arrange
    const el = <Notification message="" open={true} variant="info" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find(Notification).html()).toBe('');
  });

  it('should render an `info` notification', () => {
    // Arrange
    const el = <Notification message="Allo" open={true} variant="info" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find('div.makeStyles-info-10').exists()).toBe(true);
  });

  it('should render an `error` notification', () => {
    // Arrange
    const el = <Notification message="Allo" open={true} variant="error" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find('div.makeStyles-error-9').exists()).toBe(true);
  });

  it('should render an `success` notification', () => {
    // Arrange
    const el = <Notification message="Allo" open={true} variant="success" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find('div.makeStyles-success-8').exists()).toBe(true);
  });

  it('should render an `warning` notification', () => {
    // Arrange
    const el = <Notification message="Allo" open={true} variant="warning" />;

    // Act
    const wrapper = mount(el);

    // Assert
    expect(wrapper.find('div.makeStyles-warning-11').exists()).toBe(true);
  });
});
