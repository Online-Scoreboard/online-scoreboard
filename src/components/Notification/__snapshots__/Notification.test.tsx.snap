// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification should render a notification component when opened 1`] = `"<div class=\\"MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter\\"><div class=\\"MuiTypography-root MuiPaper-root MuiPaper-elevation6 MuiSnackbarContent-root makeStyles-info-10 MuiTypography-body2\\" role=\\"alertdialog\\" aria-describedby=\\"client-snackbar\\" style=\\"transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"><div class=\\"MuiSnackbarContent-message\\"><span id=\\"client-snackbar\\" class=\\"makeStyles-message-14\\"><svg class=\\"MuiSvgIcon-root makeStyles-iconVariant-13\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\"></path></svg>Allo</span></div></div></div>"`;
